### PR TITLE
API Server Foundation

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#96 

# Overview

- [x] I have define a new module `Cardano.Wallet.Api.Server` containing handlers and servant server foundation to implement the API
- [x] I have moved some types from 'Cardano.Wallet.Primitive.Model' to 'Cardano.Wallet.Primitive.Types'
- [x] I have (partially) implemented the `getWallet` and `postWallet` handlers from the API, leveraging what's already available in the wallet layer
- [x] I have slightly reviewed the signatures of methods in the wallet layer to make them less polymorphic (removed the monad parameter, turned out to not be useful since everything runs in IO anyway. So, I've kept the parameterization actually focused on what matters --> the wallet state)
- [x] I have added the `WalletMetadata` to the return type of `getWallet` in the wallet layer, but haven't implemented it so far (relying on lazyness here to not break everything). I'll do in a next PR.

# Comments

### CODE REVIEW

File: src/Cardano/Wallet.hs [hunk 1]
```diff
@@ -49,10 +51,16 @@ import GHC.Generics
...
+data WalletLayer s = WalletLayer
+    { createWallet
+        :: NewWallet
+        -> ExceptT CreateWalletError IO WalletId
+    , readWallet
+        :: WalletId
+        -> ExceptT ReadWalletError IO (Wallet s, WalletMetadata)
+    , watchWallet
+        :: WalletId
+        -> IO ()
```
Comment (KtorZ) [inline src/Cardano/Wallet.hs:63]: So I've made this less polymorphic to make compilation simpler and error more friendly. In the end, we do not care about the monad parameter. And, if we ever need to test that on something else than IO (like `STM`), then, we'll do it at this moment.

File: src/Cardano/Wallet.hs [hunk 2]
```diff
@@ -95,23 +103,25 @@ mkWalletLayer db network = WalletLayer
...
             Just _ ->
                 throwE $ ErrCreateWalletIdAlreadyExists wid
-    , getWallet = \wid -> lift (readCheckpoints db (PrimaryKey wid)) >>= \case
+    , readWallet = \wid -> liftIO (readCheckpoints db (PrimaryKey wid)) >>= \case
         Nothing ->
-            throwE $ ErrGetWalletNotFound wid
+            throwE $ ErrReadWalletNotFound wid
         Just (w :| _) ->
-            return w
+            return (w, error "FIXME: store and retrieve wallet metadata")
```
Comment (KtorZ) [inline src/Cardano/Wallet.hs:122]: Will do in another PR. Thanks for Haskell lazyness magic, tests still pass :sweat_smile:  ...

File: src/Cardano/Wallet/Api/Server.hs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            }
+        , delegation =
+            ApiT $ ApiT <$> meta ^. #delegation
+        , name =
+            ApiT $ meta ^. #name
+        , passphrase =
+            ApiT $ meta ^. #passphraseInfo
+        , state =
+            ApiT $ meta ^. #status
+        }
```
Comment (KtorZ) [inline src/Cardano/Wallet/Api/Server.hs:108]: I've introduced lenses here to avoid having to put qualified imports on either the primitive or the api types since we start having name clashes a bit everywhere at this level.

File: src/Cardano/Wallet/Api/Types.hs [hunk 4]
```diff
@@ -28,31 +28,21 @@
...
-    -- * Re-Export From Primitives
-    , AddressState (..)
-    , PoolId (..)
-    , WalletDelegation (..)
-    , WalletId (..)
-    , WalletName (..)
-    , WalletPassphraseInfo (..)
-    , WalletState (..)
-    , AddressPoolGap
-    , Passphrase(..)
```
Comment (KtorZ) [inline src/Cardano/Wallet/Api/Types.hs:47]: removed the re-export. Consumer modules just have to import the `Cardano.Wallet.Primitive.Types` :+1:

File: src/Cardano/Wallet/Api/Types.hs [hunk 5]
```diff
@@ -62,9 +52,18 @@ import Cardano.Wallet.Primitive.AddressDerivation
...
 import Cardano.Wallet.Primitive.Mnemonic
-import Cardano.Wallet.Primitive.Model
-    ( AddressState (..)
+    ( CheckSumBits
+    , ConsistentEntropy
+    , EntropySize
+    , entropyToBytes
+    , mkMnemonic
+    , mnemonicToEntropy
+    )
```
Comment (KtorZ) [inline src/Cardano/Wallet/Api/Types.hs:61]: This was imported implicitely :x

File: src/Cardano/Wallet/Api/Types.hs [hunk 6]
```diff
@@ -113,41 +107,36 @@ import qualified Data.Text.Encoding as T
...
+    { addressPoolGap :: !(Maybe (ApiT AddressPoolGap))
+    , mnemonicSentence :: !(ApiMnemonicT '[15,18,21,24] "seed")
+    , mnemonicSecondFactor :: !(Maybe (ApiMnemonicT '[9,12] "generation"))
+    , name :: !(ApiT WalletName)
+    , passphrase :: !(ApiT (Passphrase "encryption"))
     } deriving (Eq, Generic, Show)
 
 newtype WalletPutData = WalletPutData
-    { _name :: (Maybe (ApiT WalletName))
+    { name :: (Maybe (ApiT WalletName))
```
Comment (KtorZ) [inline src/Cardano/Wallet/Api/Types.hs:134]: removed the `_` prefix and renamed `Wallet` -> `ApiWallet` and `Address` -> `ApiAddress` to avoid confusion with primitive types.

File: src/Cardano/Wallet/Primitive/AddressDiscovery.hs [hunk 7]
```diff
@@ -226,9 +225,13 @@ nextAddresses !key (AddressPoolGap !g) !cc !fromIx =
         (>= fromIx)
     newAddress = keyToAddress . deriveAddressPublicKey key cc
 
-newtype SeqState = SeqState (AddressPool, AddressPool)
+data SeqState = SeqState
+    { internalPool :: !AddressPool
+    , externalPool :: !AddressPool
+    }
```
Comment (KtorZ) [inline src/Cardano/Wallet/Primitive/AddressDiscovery.hs:231]: Defining named accessors to prevent mistake when dealing with the state.

File: src/Cardano/Wallet/Primitive/Model.hs [hunk 8]
```diff
@@ -25,41 +24,23 @@
...
-    , WalletId(..)
-    , WalletName(..)
-    , mkWalletName
-    , walletNameMinLength
-    , walletNameMaxLength
-    , WalletNameError(..)
-    , WalletState(..)
-    , WalletDelegation (..)
-    , WalletPassphraseInfo(..)
-    , PoolId(..)
```
Comment (KtorZ) [inline src/Cardano/Wallet/Primitive/Model.hs:54]: Moved into `Cardano.Wallet.Primitive.Types`

**Comment by paweljakubas** [comment]
the compilation error is strange : 
```
src/Cardano/Wallet/Api/Server.hs:16:1: error:
    parse error on input ‘import’
   |
16 | import Prelude
```

File: src/Cardano/Wallet.hs (hunk 1 above)
Comment (paweljakubas) [reply src/Cardano/Wallet.hs:63]: :+1:

File: src/Cardano/Wallet.hs (hunk 2 above)
Comment (paweljakubas) [reply src/Cardano/Wallet.hs:122]: :+1:

File: src/Cardano/Wallet/Api/Server.hs [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    ) where
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- API handlers and server using the underlying wallet layer to provide
+-- endpoints reachable through HTTP.
+
+import Prelude
```
Comment (paweljakubas) [inline src/Cardano/Wallet/Api/Server.hs:16]: how this can not compile ?

File: src/Cardano/Wallet/Primitive/AddressDiscovery.hs (hunk 7 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/Primitive/AddressDiscovery.hs:231]: :100:

File: src/Cardano/Wallet/Primitive/Model.hs [hunk 10]
```diff
@@ -173,31 +133,29 @@ applyBlock !b (cp@(Wallet !utxo !pending _ _) :| checkpoints) =
...
-    tip
+currentTip (Wallet _ _ tip _) = tip
 
+-- | Get the wallet current state
+getState :: Wallet s -> s
+getState (Wallet _ _ _ s) = s
 
 -- | Available balance = 'balance' . 'availableUTxO'
-availableBalance :: Wallet s -> Integer
+availableBalance :: Wallet s -> Natural
```
Comment (paweljakubas) [inline src/Cardano/Wallet/Primitive/Model.hs:145]: yeah, balance cannot be negative and Natural embrace it!

**Review by paweljakubas — APPROVED** [review]
LGTM modulo not compiling because of the cryptic reason

File: src/Cardano/Wallet/Api/Server.hs (hunk 9 above)
Comment (KtorZ) [reply src/Cardano/Wallet/Api/Server.hs:16]: :thinking:  Good question? Probably `haddock` that messes up things.
